### PR TITLE
Proposed for output descriptors change, adds additional information f…

### PR DESCRIPTION
**Pull request by michaelcfanning**
…or override scenario. @lgolding  for discussion purposes

### CODE REVIEW

File: src/Sarif/Schemata/sarif-schema.json [hunk 1]
```diff
@@ -1616,39 +1616,78 @@
...
           "maximum": 100.0
         },
 
         "parameters": {
-          "description": "Contains configuration information specific to this rule.",
+          "description": "Contains configuration information specific to a notification.",
           "$ref": "#/definitions/propertyBag"
         },
 
         "properties": {
```
Comment (lgolding) [inline src/Sarif/Schemata/sarif-schema.json]: >properties [](start = 9, length = 10)

Why does an `outputConfiguration` need a property bag? (aside from its `parameters`)

File: src/Sarif/Schemata/sarif-schema.json [hunk 2]
```diff
@@ -1616,39 +1616,78 @@
...
+        },
+
+        "notificationIndex": {
+          "description": "The index within the tool component notifications descriptors array of the output descriptor associated with this override.",
+          "type": "integer",
+          "default": -1,
+          "minimum": -1
+        },
+
+        "ruleIndex": {
```
Comment (lgolding) [inline src/Sarif/Schemata/sarif-schema.json]: >ruleIndex [](start = 9, length = 9)

Ok, exactly one of `notificationIndex` and `ruleIndex` must be present.

File: src/Sarif/Schemata/sarif-schema.json [hunk 3]
```diff
@@ -2102,25 +2141,25 @@
           }
         },
 
-        "notificationsMetadata": {
+        "notificationDescriptors": {
           "description": "An array of message descriptor objects relevant to the notifications related to the configuration and runtime execution of the component.",
```
Comment (lgolding) [inline src/Sarif/Schemata/sarif-schema.json]: >message descripto [](start = 38, length = 17)

"message descriptor" =>"output descriptor" (assuming we keep the name `outputDescriptor`, which I'm struggling with).

Comment (lgolding) [reply src/Sarif/Schemata/sarif-schema.json]: Here and on `ruleDescriptors.description` below.

---
In reply to: [256190665](https://github.com/Microsoft/sarif-sdk/pull/1280#discussion_r256190665) [](ancestors = 256190665)

File: src/Sarif/Schemata/sarif-schema.json [hunk 4]
```diff
@@ -717,13 +717,13 @@
           "type": "integer"
         },
 
-        "ruleConfiguration": {
-          "description": "An array of rule configuration objects that describe runtime analysis behavior.",
+        "outputConfigurationOverrides": {
+          "description": "An array of notification configuration override objects that describe runtime reporting behavior.",
```
Comment (lgolding) [inline src/Sarif/Schemata/sarif-schema.json]: >notification [](start = 38, length = 12)

"notification configuration" => "output configuration" (assuming we keep that name).

Comment (lgolding) [reply src/Sarif/Schemata/sarif-schema.json]: Hah, you used a good word in the description! What do you think of `reportingDescriptor`, `reportingConfiguration`, and `reportingConfigurationOverride`?

---
In reply to: [256191271](https://github.com/Microsoft/sarif-sdk/pull/1280#discussion_r256191271) [](ancestors = 256191271)

Comment (lgolding) [reply src/Sarif/Schemata/sarif-schema.json]: I really like those names. Just letting you know it's not just a quibble. I really didn't like "outputXxx", and I jumped when I saw the word "reporting".

---
In reply to: [256191513](https://github.com/Microsoft/sarif-sdk/pull/1280#discussion_r256191513) [](ancestors = 256191513,256191271)

File: src/Sarif/Autogenerated/FailureLevel.cs [hunk 5]
```diff
@@ -6,7 +6,7 @@
 namespace Microsoft.CodeAnalysis.Sarif
 {
     /// <summary>
-    /// A value that specifies the failure level of the result
```
Comment (lgolding) [inline src/Sarif/Autogenerated/FailureLevel.cs]: >failure [](start = 35, length = 7)

Why remove the word "failure"? If we think the type name `FailureLevel` is descriptive (and I do), then I think it's fair to specify that this is a "failure level" (as opposed to, for example, a confidence level).

File: src/Sarif/Schemata/sarif-schema.json (hunk 4 above)
Comment (michaelcfanning) [reply src/Sarif/Schemata/sarif-schema.json]: great, i like it too. reportingDescriptor it is. will push an update.

---
In reply to: [256191779](https://github.com/Microsoft/sarif-sdk/pull/1280#discussion_r256191779) [](ancestors = 256191779,256191513,256191271)

File: src/Sarif/Autogenerated/FailureLevel.cs (hunk 5 above)
Comment (michaelcfanning) [reply src/Sarif/Autogenerated/FailureLevel.cs]: will fix, agree

---
In reply to: [256192667](https://github.com/Microsoft/sarif-sdk/pull/1280#discussion_r256192667) [](ancestors = 256192667)

**Comment by michaelcfanning** [comment]
@lgolding all tests now passing. schema reflects offline discussion.

File: src/Sarif/CodeGenHints.json [hunk 6]
```diff
@@ -432,6 +402,91 @@
...
+        ]
+      }
+    }
+  ],
+  "ReportingConfiguration.Level": [
+    {
+      "kind": "EnumHint",
+      "arguments": {
+        "typeName": "FailureLevel",
+        "description": "A value that specifies the failure level of the result",
```
Comment (lgolding) [inline src/Sarif/CodeGenHints.json]: >result [](start = 72, length = 6)

"result or notification"

File: src/ReleaseHistory.md [hunk 7]
```diff
@@ -233,9 +233,9 @@
 * API NON-BREAKING: Add 'run.newlineSequences' to schema. https://github.com/oasis-tcs/sarif-spec/issues/169
 * API NON-BREAKING: Add 'rule.deprecatedIds' to schema. https://github.com/oasis-tcs/sarif-spec/issues/293
 * API NON-BREAKING: Add 'versionControlDetails.mappedTo'. https://github.com/oasis-tcs/sarif-spec/issues/248
-* API NON-BREAKING: Add result.rank'. Add 'ruleConfiguration.defaultRank'.
+* API NON-BREAKING: Add result.rank'. Add 'reportingConfiguration.defaultRank'.
```
Comment (lgolding) [inline src/ReleaseHistory.md]: >reportingConfiguration [](start = 43, length = 22)

This should be reverted. Nothing we do in the 01-24 version affects what we published in the 01-09 version.

File: src/Sarif/Errors.cs [hunk 8]
```diff
@@ -19,7 +19,7 @@ public static class Errors
         private const string ERR997_ExceptionLoadingPlugIn = "ERR997.ExceptionLoadingPlugIn";
         private const string ERR997_NoValidAnalysisTargets = "ERR997.NoValidAnalysisTargets";
         private const string ERR997_ExceptionAccessingFile = "ERR997.ExceptionAccessingFile";
-        private const string ERR997_MissingRuleConfiguration = "ERR997.MissingRuleConfiguration";
+        private const string ERR997_MissingreportingConfiguration = "ERR997.MissingreportingConfiguration";
```
Comment (lgolding) [inline src/Sarif/Errors.cs]: >reporting [](start = 43, length = 9)

"reporting" => "Reporting"

File: src/Sarif/Baseline/ResultMatching/SarifLogMatcher.cs [hunk 9]
```diff
@@ -238,7 +238,7 @@ private MessageDescriptor GetRuleFromResources(Result result, IDictionary<string
                 properties = currentRuns.Last().Properties;
             }
 
-            var rulesMetadata = new Dictionary<MessageDescriptor, int>(MessageDescriptor.ValueComparer);
+            var ruleDescriptors = new Dictionary<ReportingDescriptor, int>(ReportingDescriptor.ValueComparer);
```
Comment (lgolding) [inline src/Sarif/Baseline/ResultMatching/SarifLogMatcher.cs]: >ruleDescriptors [](start = 16, length = 15)

`ruleDescriptors` => `reportingDescriptors`

File: src/Sarif/VersionOne/Autogenerated/IRuleVersionOneVersionOne.cs [hunk 10]
```diff
@@ -42,7 +42,7 @@ public partial interface IRuleVersionOneVersionOne
         /// <summary>
         /// A value specifying whether a rule is enabled.
         /// </summary>
-        RuleConfigurationVersionOne Configuration { get; }
```
Comment (lgolding) [inline src/Sarif/VersionOne/Autogenerated/IRuleVersionOneVersionOne.cs]: >RuleConfigurationVersionOne [](start = 8, length = 27)

No need to change any of these. There was no "reporting configuration" in version 1.

File: src/Sarif/Visitors/SarifTransformerUtilities.cs [hunk 11]
```diff
@@ -86,7 +86,7 @@ public static NotificationLevelVersionOne CreateNotificationLevelVersionOne(Fail
             }
         }
 
-        public static FailureLevel CreateRuleConfigurationDefaultLevel(ResultLevelVersionOne v1ResultLevel)
+        public static FailureLevel CreatereportingConfigurationDefaultLevel(ResultLevelVersionOne v1ResultLevel)
```
Comment (lgolding) [inline src/Sarif/Visitors/SarifTransformerUtilities.cs]: >r [](start = 41, length = 1)

r => R.

File: src/Sarif/Core/ToolComponent.cs [hunk 12]
```diff
@@ -10,14 +10,14 @@ namespace Microsoft.CodeAnalysis.Sarif
...
     public partial class ToolComponent
     {        
-        public bool ShouldSerializeRulesMetadata()
+        public bool ShouldSerializeRuleDescriptors()
         {
-            return this.RulesMetadata.HasAtLeastOneNonNullValue();
+            return this.RuleDescriptors.HasAtLeastOneNonNullValue();
         }
 
         public bool ShouldSerializeNotificationsMetadata()
```
Comment (lgolding) [inline src/Sarif/Core/ToolComponent.cs]: >ShouldSerializeNotificationsMetadata [](start = 20, length = 36)

For consistency: `ShouldSerializeNotificationDescriptors`

File: src/Sarif.Multitool/Rules/SarifValidationSkimmerBase.cs [hunk 13]
```diff
@@ -484,9 +484,9 @@ private void Visit(Message message, string messagePointer)
             Analyze(message, messagePointer);
         }
 
-        private void VisitMessageDescriptor(MessageDescriptor messageDescriptor, string messageDescriptorPointer)
+        private void VisitMessageDescriptor(ReportingDescriptor reportingDescriptor, string messageDescriptorPointer)
         {
-            Analyze(messageDescriptor, messageDescriptorPointer);
+            Analyze(reportingDescriptor, messageDescriptorPointer);
```
Comment (lgolding) [inline src/Sarif.Multitool/Rules/SarifValidationSkimmerBase.cs]: >messageDescriptorPointer [](start = 41, length = 24)

`messageDescriptorPointer` => `reportingDescriptorPointer`.

File: src/Sarif.Multitool/Rules/SarifValidationSkimmerBase.cs [hunk 14]
```diff
@@ -105,7 +105,7 @@ protected virtual void Analyze(LogicalLocation logicalLocation, string logicalLo
         protected virtual void Analyze(Message message, string messagePointer)
         {
         }
-        protected virtual void Analyze(MessageDescriptor messageDescriptor, string messageDescriptorPointer)
+        protected virtual void Analyze(ReportingDescriptor reportingDescriptor, string messageDescriptorPointer)
```
Comment (lgolding) [inline src/Sarif.Multitool/Rules/SarifValidationSkimmerBase.cs]: >messageDescriptorPointer [](start = 87, length = 24)

`messageDescriptorPointer` => `reportingDescriptorPointer`.

File: src/Sarif.FunctionalTests/DirectProducerTestData/Semmle/SarifSdk.sarif [hunk 15]
```diff
@@ -10088,7 +10088,7 @@
             }
           ]
         },
-        "#%SRCROOT%#src/Sarif/Autogenerated/RuleConfigurationEqualityComparer.cs": {
+        "#%SRCROOT%#src/Sarif/Autogenerated/reportingConfigurationEqualityComparer.cs": {
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/DirectProducerTestData/Semmle/SarifSdk.sarif]: >r [](start = 44, length = 1)

r => R

File: src/Sarif.Multitool/Rules/SarifValidationSkimmerBase.cs [hunk 16]
```diff
@@ -670,7 +670,7 @@ private void Visit(ResultProvenance resultProvenance, string resultProvenancePoi
             }
         }
 
-        private void Visit(MessageDescriptor messageDecriptor, string messageDescriptorPointer)
+        private void Visit(ReportingDescriptor messageDecriptor, string messageDescriptorPointer)
```
Comment (lgolding) [inline src/Sarif.Multitool/Rules/SarifValidationSkimmerBase.cs]: >messageDecriptor [](start = 47, length = 16)

same renames as above.

File: src/Sarif.Multitool/Rules/UriMustBeAbsolute.cs [hunk 17]
```diff
@@ -48,9 +48,9 @@ protected override void Analyze(Result result, string resultPointer)
             }
         }
 
-        protected override void Analyze(MessageDescriptor messageDescriptor, string messageDescriptorPointer)
+        protected override void Analyze(ReportingDescriptor reportingDescriptor, string messageDescriptorPointer)
         {
-            AnalyzeUri(messageDescriptor.HelpUri, messageDescriptorPointer.AtProperty(SarifPropertyName.HelpUri));
+            AnalyzeUri(reportingDescriptor.HelpUri, messageDescriptorPointer.AtProperty(SarifPropertyName.HelpUri));
```
Comment (lgolding) [inline src/Sarif.Multitool/Rules/UriMustBeAbsolute.cs]: >messageDescriptorPointer [](start = 52, length = 24)

`reportingDescriptorPointer`

File: src/Sarif.Multitool/Rules/UrisMustBeValid.cs [hunk 18]
```diff
@@ -53,9 +53,9 @@ protected override void Analyze(Result result, string resultPointer)
             }
         }
 
-        protected override void Analyze(MessageDescriptor messageDescriptor, string messageDescriptorPointer)
+        protected override void Analyze(ReportingDescriptor reportingDescriptor, string messageDescriptorPointer)
         {
-            AnalyzeUri(messageDescriptor.HelpUri, messageDescriptorPointer.AtProperty(SarifPropertyName.HelpUri));
+            AnalyzeUri(reportingDescriptor.HelpUri, messageDescriptorPointer.AtProperty(SarifPropertyName.HelpUri));
```
Comment (lgolding) [inline src/Sarif.Multitool/Rules/UrisMustBeValid.cs]: >messageDescriptorPointer [](start = 52, length = 24)

`reportingDescriptorPointer`

File: src/Sarif.UnitTests/Writers/SarifLoggerTests.cs [hunk 19]
```diff
@@ -328,7 +328,7 @@ public void SarifLogger_ScrapesFilesFromResult()
                     invocationPropertiesToLog: null))
                 {
                     string ruleId = "RuleId";
-                    var rule = new MessageDescriptor() { Id = ruleId };
+                    var rule = new ReportingDescriptor() { Id = ruleId };
```
Comment (lgolding) [inline src/Sarif.UnitTests/Writers/SarifLoggerTests.cs]: >()  [](start = 54, length = 3)

parens unnecessary

**Comment by lgolding** [comment]
Pushed changes for feedback. Left a few as a tail since you are pressed for time.

**Comment by lgolding** [comment]
>         "driver": {

@michaelcfanning I was updating the SDK to conform to our agreement _not_ to replace `conversion.tool` with `conversion.driver`. So I changed this property from `driver` to `tool`, and all of a sudden five tests started failing, complaining that the test file did not contain the required property `runs[0].conversion.driver`.

It turns out that those test files had _never_ contained `runs[0].conversion.driver`. So why did the tests pass after your toolComponent/notification metadata changes, but started failing when I replaced `driver` with `tool` at this point in the schema?

It's because in my review, I missed this bug: when you replaced `conversion.tool` with `conversion.driver` on L123, you did _not_ replace `tool` with `driver` in the list of required properties on L151. As a result, the JSON serializer allowed these files to continue to have a `conversion.tool` property, but it did not realize that the `conversion.tool` was supposed to be a `tool` object -- and so it did not complain when those `conversion.tool` properties didn't have a `driver` sub-property!

In my current PR, I'm re-replacing `conversion.driver` (of type `toolComponent`) with `conversion.tool` (of type `tool`), so the schema definition for `conversion` will once again be internally consistent. That exposes our previous failure to update the test files for the failing tests, so I'll fix those as well.

---
Refers to: src/Sarif/Schemata/sarif-schema.json:123 in e5bee9d. [](commit_id = e5bee9df2cc96d46d177a9a14284bb12844c6a35, deletion_comment = False)

**Pull request merged**